idb/postgres/internal/writer/writer.go

```diff
@@ -364,7 +379,35 @@ func (w *Writer) addTransactionParticipation(block *bookkeeping.Block) error {
 	return nil
 }
 
-func writeAccountData(round basics.Round, address basics.Address, accountData basics.AccountData, batch *pgx.Batch) {
+type sigTypeDelta struct {
+	present bool
+	value   idb.SigType
+}
+
+func getSigTypeDeltas(payset []transactions.SignedTxnInBlock) (map[basics.Address]sigTypeDelta, error) {
+	res := make(map[basics.Address]sigTypeDelta, len(payset))
+
+	for i := range payset {
+		if payset[i].Txn.RekeyTo == (basics.Address{}) {
+			sigtype, err := idb.SignatureType(&payset[i].SignedTxn)
+			if err != nil {
+				return nil, fmt.Errorf("getSigTypeDelta() err: %w", err)
+			}
+			res[payset[i].Txn.Sender] = sigTypeDelta{present: true, value: sigtype}
+		} else {
+			res[payset[i].Txn.Sender] = sigTypeDelta{}
+		}
+	}
+
+	return res, nil
+}
+
+type optionalSigTypeDelta struct {
```

Comment:
This nested optional implementation is really confusing. Even knowing the problem it solves I still had a lot of trouble following how it works.

I think the code would be easier to understand with a single optional type (the one named `sigTypeDelta`), and a check for `value.IsZero()`. That way you either set the value to zero, or skip setting it.



 is more complicated than the `writeAccounts` function you replaced with it. I think it would be better to have a single optional and check `value.IsZero()` for

Reply:
The reason why I did it this way is because `idb.SigType` doesn't have a special zero value. I think it shouldn't. I will add a comment for `sigTypeDelta`.